[BUG] [MCP] SSE channel failure due to socket read timeout

Getting SSE channel failure due to socket read timeout after some interval ( ideal after 5 minutes)

**Log and Stack trace**
`transport.http.SseEventListener  : SSE channel failure

java.net.SocketTimeoutException: timeout
        at okio.SocketAsyncTimeout.newTimeoutException(JvmOkio.kt:146) ~[okio-jvm-3.6.0.jar:na]
        at okio.AsyncTimeout.access$newTimeoutException(AsyncTimeout.kt:161) ~[okio-jvm-3.6.0.jar:na]
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:339) ~[okio-jvm-3.6.0.jar:na]     
        at okio.RealBufferedSource.request(RealBufferedSource.kt:209) ~[okio-jvm-3.6.0.jar:na]
        at okio.RealBufferedSource.require(RealBufferedSource.kt:202) ~[okio-jvm-3.6.0.jar:na]
        at okio.RealBufferedSource.readHexadecimalUnsignedLong(RealBufferedSource.kt:384) ~[okio-jvm-3.6.0.jar:na]
        at okhttp3.internal.http1.Http1ExchangeCodec$ChunkedSource.readChunkSize(Http1ExchangeCodec.kt:437) ~[okhttp-4.12.0.jar:na]
        at okhttp3.internal.http1.Http1ExchangeCodec$ChunkedSource.read(Http1ExchangeCodec.kt:416) ~[okhttp-4.12.0.jar:na]
        at okhttp3.internal.connection.Exchange$ResponseBodySource.read(Exchange.kt:281) ~[okhttp-4.12.0.jar:na]
        at okio.RealBufferedSource.select(RealBufferedSource.kt:232) ~[okio-jvm-3.6.0.jar:na]
        at okhttp3.internal.sse.ServerSentEventReader.processNextEvent(ServerSentEventReader.kt:50) ~[okhttp-sse-4.12.0.jar:na]
        at okhttp3.internal.sse.RealEventSource.processResponse(RealEventSource.kt:75) ~[okhttp-sse-4.12.0.jar:na]
        at okhttp3.internal.sse.RealEventSource.onResponse(RealEventSource.kt:46) ~[okhttp-sse-4.12.0.jar:na]
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519) ~[okhttp-4.12.0.jar:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.base/sun.nio.ch.NioSocketImpl.timedRead(NioSocketImpl.java:278) ~[na:na]     
        at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:304) ~[na:na]      
        at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:346) ~[na:na]
        at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:796) ~[na:na]        
        at java.base/java.net.Socket$SocketInputStream.read(Socket.java:1099) ~[na:na]       
        at okio.InputStreamSource.read(JvmOkio.kt:93) ~[okio-jvm-3.6.0.jar:na]
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:128) ~[okio-jvm-3.6.0.jar:na]     
        ... 14 common frames omitted`

**To Reproduce**
MCP client server repositories:
https://github.com/RamAlapure/fintech-mcp-server
https://github.com/RamAlapure/fintech-mcp-client 

**Expected behavior**
Need something to recover from the socket timeout

**Please complete the following information:**
- LangChain4j version: 1.0.0-beta3 (MCP client)
- LLM(s) used: OpenAI gpt-4o-mini
- Java version: 21
- Spring Boot version (if applicable): 3.4.4
- Spring AI version: 1.0.0-M7 (MCP server)

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/DefaultMcpClient.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/DefaultMcpClient.java
@@ -64,6 +64,7 @@ public class DefaultMcpClient implements McpClient {
     private final AtomicReference<List<ToolSpecification>> toolListRefs = new AtomicReference<>();
     private final AtomicBoolean toolListOutOfDate = new AtomicBoolean(true);
     private final AtomicReference<CompletableFuture<Void>> toolListUpdateInProgress = new AtomicReference<>(null);
+    private final Duration reconnectInterval;
 
     public DefaultMcpClient(Builder builder) {
         transport = ensureNotNull(builder.transport, "transport");
@@ -75,6 +76,7 @@ public class DefaultMcpClient implements McpClient {
         promptsTimeout = getOrDefault(builder.promptsTimeout, Duration.ofSeconds(60));
         logHandler = getOrDefault(builder.logHandler, new DefaultMcpLogMessageHandler());
         pingTimeout = getOrDefault(builder.pingTimeout, Duration.ofSeconds(10));
+        reconnectInterval = getOrDefault(builder.reconnectInterval, Duration.ofSeconds(5));
         toolExecutionTimeoutErrorMessage =
                 getOrDefault(builder.toolExecutionTimeoutErrorMessage, "There was a timeout executing the tool");
         RESULT_TIMEOUT = JsonNodeFactory.instance.objectNode();
@@ -86,6 +88,15 @@ public class DefaultMcpClient implements McpClient {
                 .addObject()
                 .put("type", "text")
                 .put("text", toolExecutionTimeoutErrorMessage);
+        transport.onFailure(() -> {
+            try {
+                TimeUnit.MILLISECONDS.sleep(reconnectInterval.toMillis());
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            log.info("Trying to reconnect...");
+            initialize();
+        });
         initialize();
     }
 
@@ -346,6 +357,7 @@ public class DefaultMcpClient implements McpClient {
         private Duration pingTimeout;
         private Duration promptsTimeout;
         private McpLogMessageHandler logHandler;
+        private Duration reconnectInterval;
 
         public Builder transport(McpTransport transport) {
             this.transport = transport;
@@ -439,6 +451,15 @@ public class DefaultMcpClient implements McpClient {
          */
         public Builder pingTimeout(Duration pingTimeout) {
             this.pingTimeout = pingTimeout;
+            return this;
+        }
+
+        /**
+         * The delay before attempting to reconnect after a failed connection.
+         * The default is 5 seconds.
+         */
+        public Builder reconnectInterval(Duration reconnectInterval) {
+            this.reconnectInterval = reconnectInterval;
             return this;
         }
 

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/McpTransport.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/McpTransport.java
@@ -42,4 +42,6 @@ public interface McpTransport extends Closeable {
      * that the server subprocess isn't alive anymore.
      */
     void checkHealth();
+
+    void onFailure(Runnable actionOnFailure);
 }

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/HttpMcpTransport.java
@@ -36,6 +36,7 @@ public class HttpMcpTransport implements McpTransport {
     private final boolean logRequests;
     private EventSource mcpSseEventListener;
     private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private volatile Runnable onFailure;
 
     // this is obtained from the server after initializing the SSE channel
     private volatile String postUrl;
@@ -104,6 +105,11 @@ public class HttpMcpTransport implements McpTransport {
         // no transport-specific checks right now
     }
 
+    @Override
+    public void onFailure(Runnable actionOnFailure) {
+        this.onFailure = actionOnFailure;
+    }
+
     private CompletableFuture<JsonNode> execute(Request request, Long id) {
         CompletableFuture<JsonNode> future = new CompletableFuture<>();
         if (id != null) {
@@ -137,7 +143,8 @@ public class HttpMcpTransport implements McpTransport {
     private EventSource startSseChannel(boolean logResponses) {
         Request request = new Request.Builder().url(sseUrl).build();
         CompletableFuture<String> initializationFinished = new CompletableFuture<>();
-        SseEventListener listener = new SseEventListener(messageHandler, logResponses, initializationFinished);
+        SseEventListener listener =
+                new SseEventListener(messageHandler, logResponses, initializationFinished, onFailure);
         EventSource eventSource = EventSources.createFactory(client).newEventSource(request, listener);
         // wait for the SSE channel to be created, receive the POST url from the server, throw an exception if that
         // failed

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/SseEventListener.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/SseEventListener.java
@@ -19,12 +19,17 @@ public class SseEventListener extends EventSourceListener {
     // this will contain the POST url for sending commands to the server
     private final CompletableFuture<String> initializationFinished;
     private final McpOperationHandler messageHandler;
+    private final Runnable onFailure;
 
     public SseEventListener(
-            McpOperationHandler messageHandler, boolean logEvents, CompletableFuture initializationFinished) {
+            McpOperationHandler messageHandler,
+            boolean logEvents,
+            CompletableFuture initializationFinished,
+            Runnable onFailure) {
         this.messageHandler = messageHandler;
         this.logEvents = logEvents;
         this.initializationFinished = initializationFinished;
+        this.onFailure = onFailure;
     }
 
     @Override
@@ -65,6 +70,9 @@ public class SseEventListener extends EventSourceListener {
         }
         if (t != null && (t.getMessage() == null || !t.getMessage().contains("Socket closed"))) {
             log.warn("SSE channel failure", t);
+            if (onFailure != null) {
+                onFailure.run();
+            }
         }
     }
 

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdio/StdioMcpTransport.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdio/StdioMcpTransport.java
@@ -94,6 +94,11 @@ public class StdioMcpTransport implements McpTransport {
     }
 
     @Override
+    public void onFailure(Runnable actionOnFailure) {
+        // ignore, for stdio transport, we currently don't do reconnection attempts
+    }
+
+    @Override
     public void close() throws IOException {
         process.destroy();
     }

--- a/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpReconnectIT.java
+++ b/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpReconnectIT.java
@@ -1,0 +1,68 @@
+package dev.langchain4j.mcp.client.integration;
+
+import static dev.langchain4j.mcp.client.integration.McpServerHelper.skipTestsIfJbangNotAvailable;
+import static dev.langchain4j.mcp.client.integration.McpServerHelper.startServerHttp;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import dev.langchain4j.mcp.client.DefaultMcpClient;
+import dev.langchain4j.mcp.client.transport.McpTransport;
+import dev.langchain4j.mcp.client.transport.http.HttpMcpTransport;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.TimeoutException;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class McpReconnectIT {
+
+    private static Process process;
+    private static DefaultMcpClient mcpClient;
+
+    @BeforeAll
+    static void setup() throws IOException, InterruptedException, TimeoutException {
+        skipTestsIfJbangNotAvailable();
+        process = startServerHttp("tools_mcp_server.java");
+        McpTransport transport = new HttpMcpTransport.Builder()
+                .sseUrl("http://localhost:8080/mcp/sse")
+                .logRequests(true)
+                .logResponses(true)
+                .build();
+        mcpClient = new DefaultMcpClient.Builder()
+                .transport(transport)
+                .toolExecutionTimeout(Duration.ofSeconds(4))
+                .reconnectInterval(Duration.ofSeconds(1))
+                .build();
+    }
+
+    @Test
+    public void testReconnect() throws IOException, TimeoutException, InterruptedException {
+        executeAToolAndAssertSuccess();
+
+        // kill the server and restart it
+        process.destroy();
+        process = startServerHttp("tools_mcp_server.java");
+
+        // give the MCP client some time to reconnect
+        Thread.sleep(5_000);
+
+        executeAToolAndAssertSuccess();
+    }
+
+    @AfterAll
+    static void tearDown() throws IOException, InterruptedException {
+        if (process != null) {
+            process.destroyForcibly();
+        }
+    }
+
+    private void executeAToolAndAssertSuccess() {
+        ToolExecutionRequest toolExecutionRequest = ToolExecutionRequest.builder()
+                .name("echoString")
+                .arguments("{\"input\": \"abc\"}")
+                .build();
+        String result = mcpClient.executeTool(toolExecutionRequest);
+        assertThat(result).isEqualTo("abc");
+    }
+}

